### DISCUSSION TITLE
[varLib.instancer] adding / replacing named instances in a sliced VF

### DISCUSSION THREAD

**Discussion by JeremieHornus** [fonttools/fonttools#2817] [Ideas]
Wouldn't it be a super nice feature to add to the `instantiateVariableFont` function that one can add or replace the named instances of the output VF ? For example the extremes may lose their named instances when slicing and it would be good to be able to add new. Also the existing named instance may not make sense anymore so replacing them with new values/names would be very handy.
https://github.com/fonttools/fonttools/blob/276f6aaf4480ec2ad81cbd5fcc2210af78c0c3f8/Lib/fontTools/varLib/instancer/__init__.py#L1146

I know it's probably possible to do 'manually' by instantiating then rebuilding the VF from a new designspace with the instanciated fonts but I feel the instancer would be much more powerful with this feature built in.

I'd be glad to hear opinions here.

**Comment by anthrotype** [comment]
What do you suggest as an interface? Can you provide an example?
Instead of rebuilding the VF from a new designspace, it might be simpler/more effective to just patch the resulting instanced font's fvar/name tables as desired

**Reply by fabiocaccamo** [reply]
I think that an interface like this would be the best you could want in terms of simplicity and effectiveness:

```python
instance_names_mapping = {
    "ExtraLight": "Custom ExtraLight",
    "Regular": "Custom Regular",
    "Medium": "Custom Medium",
    "Bold": "Custom Bold",
    "ExtraBold": "Custom ExtraBold",
    "ExtraLight Italic": "Custom ExtraLight Italic",
    "Italic": "Custom Italic",
    "Medium Italic": "Custom Medium Italic",
    "Bold Italic": "Custom Bold Italic",
    "ExtraBold Italic": "Custom ExtraBold Italic",
}

renameInstances(mySlicedVF.ttf, names=instance_names_mapping)
```

**Comment by JeremieHornus** [comment]
Indeed, one would feed the interface with a designspace-like file and the sliced VF to that the tool patches the named instances to it.
something like:
`renameInstances(mySlicedVF.ttf, myDesignSpace.designspace)`

**Reply by JeremieHornus** [reply]
or maybe a STAT table entry could be enough to deduce fvar/name?
@justvanrossum and I are working on a command-line tool to enable this quickly for now

**Reply by anthrotype** [reply]
if you can come up with something general purpose that can be upstreamed here, all for the better, thank you
